articles/concepts-grovers.md

```diff
@@ -0,0 +1,185 @@
+ms.author: v-edsanc
+ms.date: 12/11/2017
+ms.topic: article
+ms.prod: azure-quantum
+no-loc: ['Q#', '$$v', '$$', "$$", '$', "$", $, $$, '\cdots', 'bmatrix', '\ddots', '\equiv', '\sum', '\begin', '\end', '\sqrt', '\otimes', '{', '}', '\text', '\phi', '\kappa', '\psi', '\alpha', '\beta', '\gamma', '\delta', '\omega', '\bra', '\ket', '\boldone', '\\\\', '\\', '=', '\frac', '\text', '\mapsto', '\dagger', '\to', '\begin{cases}', '\end{cases}', '\operatorname', '\braket', '\id', '\expect', '\defeq', '\variance', '\dd', '&', '\begin{align}', '\end{align}', '\Lambda', '\lambda', '\Omega', '\mathrm', '\left', '\right', '\qquad', '\times', '\big', '\langle', '\rangle', '\bigg', '\Big', '|', '\mathbb', '\vec', '\in', '\texttt', '\ne', '<', '>', '\leq', '\geq', '~~', '~', '\begin{bmatrix}', '\end{bmatrix}', '\_']
+---
+
+# Theory of Grover's search algorithm
+
+In this article you'll find a detailed theoretical explanation of the
+mathematical principles that make Grover's algorithm work.
+
+For a practical implementation of Grover's algorithm to solve mathematical
+problems you can read our [guide to implement Grover's search algorithm](xref:microsoft.quantum.tutorial-qdk.grovers).
+
+## Statement of the problem
+
+Any searching task can be expressed with an abstract function $f(x)$ that accepts search items $x$. If the item $x$ is a solution for the search task, then $f(x)=1$. If the item $x$ isn't a solution, then $f(x)=0$. The search problem consists in finding any item $x_0$ such that $f(x_0)=1$. This is, an item $x_0$ that is a solution to the search problem.
+
+The task that Grover's algorithm aims to solve is, given a classical function $f(x):\\{0,1\\}^n \rightarrow\\{0,1\\}$, find an input $x_0$ for which $f(x_0)=1$.
+
+## Outline of the algorithm
+
+Suppose we have $N=2^n$ eligible items for the search task and we index them by assigning each item an integer from $0$ to
+$N-1$. The steps of the algorithm are:
+
+1. Start with a register of $n$ qubits initialized in the state $\ket{0}$ by applying $H$ to each qubit of the register.
+1. Prepare the register into a uniform superposition: $$|\text{register}\rangle=\frac{1}{N^{1 / 2}} \sum_{x=0}^{N-1}|x\rangle$$
+1. Apply $N_{\text{optimal}}$ times the following operations to the register:
+   1. The phase oracle $O_f$ that applies a conditional phase shift of $-1$ for the solution items.
+   1. Apply $H$ to each qubit of the register.
+   1. A conditional phase shift of $-1$ to every computational basis state except $\ket{0}$. This can be represented by the unitary operation $O_0$.
+   1. Apply $H$ to each qubit of the register.
+1. Measure the register to obtain the index of a item that's a solution with very high probability.
+1. Check if it's a valid solution. If not, start again.
+
+$N_\text{optimal} = \lfloor\lfloor \frac{\pi}{4}\sqrt{\frac{N}{M}}-\frac{1}{2} \right\rflor$ is the optimal number of iterations that maximizes the likelihood of obtaining the correct item by measuring the register.
```

Comment:
> $N_\text{optimal} = \lfloor\lfloor \frac{\pi}{4}\sqrt{\frac{N}{M}}-\frac{1}{2} \right\rflor$ is the optimal number of iterations that maximizes the likelihood of obtaining the correct item by measuring the register.

Here what does $M$ represent?

Reply:
Ah true, it's defined later in the tutorial. I'll introduce it earlier.

Reply:
That would be great, thanks!